src/browser-frontend/sagas/server-event-handlers.js

```diff
@@ -0,0 +1,47 @@
+/*
+Copyright 2016 Mozilla
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+this file except in compliance with the License. You may obtain a copy of the
+License at http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software distributed
+under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+*/
+
+import { takeEvery, call, put, select } from 'redux-saga/effects';
+
+import SharedActions from '../../shared/actions/shared-actions';
+import PagesModelActions from '../actions/pages-model-actions';
+import * as DomainPagesSelectors from '../selectors/domain-pages-selectors';
+import * as UIPagesSelectors from '../selectors/ui-pages-selectors';
+
+function* handleCmdOrControlQ({ meta: client }) {
+  yield call([client, client.send], SharedActions.events.fromFrontend.toServer.app.window.requestedQuit());
+}
+
+function* handleCmdOrControlW({ meta: client }) {
+  const count = (yield select(DomainPagesSelectors.getPages)).count();
+  if (count === 1) {
+    client.send(SharedActions.events.fromFrontend.toServer.app.window.requestedClose());
+    return;
+  }
+
+  const selectedPageId = yield select(UIPagesSelectors.getSelectedPageId);
+  yield put(PagesModelActions.removePage({ pageId: selectedPageId }));
+}
+
+function* onKeyShortcutPressed({ meta: client, payload: { shortcut } }) {
+  if (shortcut.keys === 'CommandOrControl+Q') {
```

Comment:
In the distant past, I had a lot of problems with long chains of `if/else if` blocks.  Maybe this one won't grow too much further, but if we add another if, I would suggest switching it to a map of strings to functions…